src/hubRepository.ts

```diff
@@ -0,0 +1,141 @@
+import * as httpm from '@actions/http-client';
+import {Index} from './types/oci';
+import os from 'os';
+import * as core from '@actions/core';
+import {Manifest} from './types/oci/manifest';
+import * as tc from '@actions/tool-cache';
+import fs from 'fs';
+import {MEDIATYPE_IMAGE_INDEX_V1, MEDIATYPE_IMAGE_MANIFEST_V1} from './types/oci/mediatype';
+import {MEDIATYPE_IMAGE_MANIFEST_V2, MEDIATYPE_IMAGE_MANIFEST_LIST_V2} from './types/docker/mediatype';
+import {DockerHub} from './dockerhub';
+
+export class HubRepository {
+  private repo: string;
+  private token: string;
+  private static readonly http: httpm.HttpClient = new httpm.HttpClient('setup-docker-action');
+
+  private constructor(repository: string, token: string) {
+    this.repo = repository;
+    this.token = token;
+  }
+
+  public static async build(repository: string): Promise<HubRepository> {
+    const token = await this.getToken(repository);
+    return new HubRepository(repository, token);
+  }
+
+  // Unpacks the image layers and returns the path to the extracted image.
+  // Only OCI indexes/manifest list are supported for now.
+  public async extractImage(tag: string, destDir?: string): Promise<string> {
```

Comment:
Wonder if we should move this to https://github.com/docker/actions-toolkit/blob/main/src/oci/oci.ts

Reply:
Ah my bad just saw OCI media types are not supported

Reply:
They are! 😅 
My point here was that it only works with images that target index, not a manifest directly.

Reply:
Not sure if it would make sense to move it - there's not much actual image parsing, most of what this function does is fetches data from Hub. But perhaps we could move the platform-specific manifest selection logic to the OCI?

Reply:
Maybe we could just side load https://crazymax.dev/undock/ and extract the image with it to make that less error prone?

Reply:
We could! I went with this approach as I assumed that we'd like to make it as minimal as possible. I think that could take a couple of seconds in worst case (fetch binary and run it)?

Reply:
That was just to avoid manifest platform check https://github.com/docker/actions-toolkit/pull/438/files#diff-0f2ffb1716f62956ee604fae748f3369d867fa5c39a60388479728f013479b46R118 so we can rely on go runtime instead :sweat_smile: Also that's pretty fast within GH infrastructure to fetch assets from GH releases.

Reply:
We'll still have a similar logic to determine which undock binary to fetch though?

Reply:
True would be similar to buildx one https://github.com/docker/actions-toolkit/blob/2eba5d9d103d5a314b8dcfada334a3cca52459ad/src/buildx/install.ts#L213-L238

Reply:
Let's keep that for a follow-up, that's fine by me!